### PR TITLE
VITIS-12947 - Decode the mask value which is used for patching a scale value

### PR DESCRIPTION
**Pull request by dezhiAmd**
#### Problem solved by the commit
Decode the mask value which is used for patching a scale value (Credit to @NishadSaraf  explain to me the local memtile address)
#### How problem was solved, alternative solutions (if any) and why they were rejected
The mask value is encoded in the elf file. Read the "value" column and store in the array associated with offset
#### What has been tested and how, request additional testing if necessary
SimNowLite test **passes** on PHX and StrixB0

### CODE REVIEW

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 1]
```diff
@@ -152,12 +154,28 @@ struct patcher
...
+    int shift_count = 0;
+
+    while ((mask & 1) == 0) {
+      mask >>= 1;
+      shift_count++;
+    }
+
+    new_value &= clear_mask;
+    new_value |= static_cast<uint32_t>(register_value << shift_count);
+    *data_to_patch = new_value;
```
Comment (larry9523) [inline src/runtime_src/core/common/api/xrt_module.cpp:178]: mask write can be as simple as
value = (value & ~mask) | (new_value & mask);

**Review by larry9523 — COMMENTED** [review]
please simplify the mask write implementation

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 2]
```diff
@@ -670,17 +683,22 @@ class module_elf : public module_impl
...
-
         std::string argnm{ symname, symname + std::min(strlen(symname), dynstr->get_size()) };
+
+        auto patch_scheme = static_cast<patcher::symbol_type>(rela->r_addend & schema_mask);
+
+        patcher::patch_info pi = patch_scheme == patcher::symbol_type::scalar_32bit_kind ?
+                                 // st_size is is encoded using register value mask for scaler_32
+                                 // for other pacthing scheme it is encoded using size of dma
+                                 patcher::patch_info{ offset, add_end_higher_28bit, static_cast<uint32_t>(sym->st_size) } :
+                                 patcher::patch_info{ offset, add_end_higher_28bit, 0 };
```
Comment (dezhiAmd) [inline src/runtime_src/core/common/api/xrt_module.cpp:694]: Check this line

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 3]
```diff
@@ -204,22 +213,26 @@ struct patcher
...
-  patch(uint8_t* base, uint64_t bo_addr)
+  patch(uint8_t* base, uint64_t new_value)
   {
     for (auto item : m_ctrlcode_patchinfo) {
       auto bd_data_ptr = reinterpret_cast<uint32_t*>(base + item.offset_to_patch_buffer);
       switch (m_symbol_type) {
       case symbol_type::scalar_32bit_kind:
-        patch32(bd_data_ptr, bo_addr + item.offset_to_base_bo_addr);
+        // new_value is a register value
+        patch32(bd_data_ptr, new_value, item.mask);
```
Comment (dezhiAmd) [inline src/runtime_src/core/common/api/xrt_module.cpp:223]: item.mask could be 0

**Review by larry9523 — APPROVED** [review]

File: src/runtime_src/core/common/api/xrt_module.cpp [hunk 4]
```diff
@@ -204,22 +211,27 @@ struct patcher
...
-  patch(uint8_t* base, uint64_t bo_addr)
+  patch(uint8_t* base, uint64_t new_value)
   {
     for (auto item : m_ctrlcode_patchinfo) {
       auto bd_data_ptr = reinterpret_cast<uint32_t*>(base + item.offset_to_patch_buffer);
       switch (m_symbol_type) {
       case symbol_type::scalar_32bit_kind:
-        patch32(bd_data_ptr, bo_addr + item.offset_to_base_bo_addr);
+        // new_value is a register value
+        if (!item.mask)
```
Comment (dezhiAmd) [inline src/runtime_src/core/common/api/xrt_module.cpp:221]: Typo here, need fix

**Pull request merged**